### PR TITLE
Describe how to use build artifact from a docu repo PR to create the webpage

### PR DESCRIPTION
**Pull request by pgfeller**
Fixes https://github.com/openhab/website/issues/455

This PR adds additional functionality to build the website using changes from a PR from the documentation repository.

Contributors:
  - @florian-h05 
  - @pgfeller 

Special Thanks to:
  - @stefan-hoehn 
----

In special situations; e.g. if you make changes in the documentation repository that might impact the website build, it is possible to build the website from a PR. This makes it possible to verify that your proposed changes do not have negative side effects:

```bash
ARGUMENTS="--pull-request 2272" npm run build-local
```
Available ARGUMENTS:

- `--verbose`
- `--no-clone`
- `--pull-request #`

### CODE REVIEW

**Comment by pgfeller** [comment]
@florian-h05 and @stefan-hoehn - please have a look at the proposed changes. This PR adds some arguments to create and serve the documentation page locally based on a PR. **It needs some fine tuning** - but I would appreciate your feedback if this adds value add all - or if the "feature" is not necessary.

```bash
ARGUMENTS="--pull-request 2272" npm run build-local

> openhab-website@3.0.0 prebuild-local
> npm run prepare-docs


> openhab-website@3.0.0 prepare-docs
> ruby prepare-docs.rb $ARGUMENTS

➡️ Cloning repository 📦 dilyanpalauzov:oh_addons_install ...
  ↪️ PR #2272: Document addon installation in Karaf from Marketplace
➡️ Migrating the introduction article
➡️ Migrating common images
➡️ Migrating logos
➡️ Migrating the Concepts section
➡️ Migrating the Installation section
➡️ Migrating the Tutorial section
➡️ Migrating the Configuration section
➡️ Migrating the Main UI section
➡️ Migrating the Migration Tutorial section
➡️ Migrating the Blockly Tutorial section
➡️ Migrating the UI section
process_file: IGNORING (NON-EXISTING): .vuepress/openhab-docs/_addons_uis/habpanel/doc/habpanel.md
process_file: IGNORING (NON-EXISTING): .vuepress/openhab-docs/_addons_uis/habot/readme.md
➡️ Migrating the Apps section
➡️ Migrating the Administration section
➡️ Migrating the Developer section
➡️ Migrating add-ons: Automation
➡️ Migrating add-ons: Persistence
➡️ Migrating add-ons: Transformations
➡️ Migrating add-ons: Voice
➡️ Migrating add-ons: IO
➡️ Migrating add-ons: UI
⚠️  Ecosystem documentation for [Alexy, Mycroft, Google] skipped - depends on Git-Workflow ...
➡️ Migrating add-ons: Bindings
➡️ Creating ZWave thing viewer
➡️ Writing add-ons arrays to files for sidebar navigation
⚠️  Iconsets depend on Git-Workflow - will be skipped ...
➡️ Downloading and extracting latest Javadoc from Jenkins
2024-03-10 23:58:12 URL:https://ci.openhab.org/job/openHAB-JavaDoc/lastSuccessfulBuild/artifact/target/javadoc-latest.tgz [4357840/4357840] -> "javadoc-latest.tgz.8" [1]
⚠️  Thingtypes depend on Git-Workflow - will be skipped ...

> openhab-website@3.0.0 build-local
> npm run dev --silent
```

image

... let me know if I shall continue - or if we abandon this, because we do not need it. In that case we might consider to salvage some parts (e.g. improved processing of arguments in the `ruby` script.

with kind regards

**Comment by stefan-hoehn** [comment]
I think that the idea definitely makes sense and I like it a lot. The more and easier we can try out changes locally the better it is. 😍

In general it looks good to me. I would recommend having a joint session for trying it out together.

@florian-h05 I would appreciate if you have a closer look at the code as well.

File: prepare-docs.rb [hunk 1]
```diff
@@ -5,38 +5,100 @@
 require "net/http"
 require "uri"
 require "rexml/document"
-# require "nokogiri"
```
Comment (stefan-hoehn) [inline prepare-docs.rb:8]: What was that for and why did you remove it?

File: prepare-docs.rb [hunk 2]
```diff
@@ -5,38 +5,100 @@
...
+
+    system("OH_DOCS_VERSION=#{branch}")
+  
+    FileUtils.cd(target_directory, verbose: false) do
+      system("git clone --depth 1 #{repository_url} --branch #{branch} #{$verbose ? '' : '--quiet'}")
+      system("git reset ##{sha} #{$verbose ? '' : '--quiet'}")
+    end
+  end
+
+verbose "🧹 Cleaning existing documentation downloads ..."
```
Comment (stefan-hoehn) [inline prepare-docs.rb:49]: I love the broom here <3

File: prepare-docs.rb [hunk 3]
```diff
@@ -681,17 +743,27 @@ def process_file(indir, file, outdir, source)
 }
 
 
-# Regenerate the classic iconset docs
-puts ">>> Generating iconset"
+# Git workflow for PRs are note executed - therefore the _addons_ios folder is not present for PR checks - this section will be scipped.
```
Comment (stefan-hoehn) [inline prepare-docs.rb:746]: Same typos like above in line 646

File: prepare-docs.rb [hunk 4]
```diff
@@ -681,17 +743,27 @@ def process_file(indir, file, outdir, source)
...
 `[ -e .vuepress/public/javadoc/latest ] && rm -r .vuepress/public/javadoc/latest`
 # Publish latest Javadoc
-puts ">>> Downloading and extracting latest Javadoc from Jenkins"
+puts "➡️ Downloading and extracting latest Javadoc from Jenkins"
 `wget -nv https://ci.openhab.org/job/openHAB-JavaDoc/lastSuccessfulBuild/artifact/target/javadoc-latest.tgz`
 `tar xzvf javadoc-latest.tgz --strip 2 && mv apidocs/ .vuepress/public/javadoc/latest`
 
-# Copy the thing-types.json file to the proper location
-FileUtils.cp(".vuepress/openhab-docs/.vuepress/thing-types.json", ".vuepress")
+# Git workflow for PRs are note executed - therefore the _addons_ios folder is not present for PR checks - this section will be scipped.
```
Comment (stefan-hoehn) [inline prepare-docs.rb:763]: Same typos like above in line 646

**Review by stefan-hoehn — CHANGES_REQUESTED** [review]

**Review by florian-h05 — COMMENTED** [review]
In general I really like this PR and I think it will prove helpful 👍 

A few first comments, I have had a look at all files except the Ruby script.
I will review the Ruby script as well once you have addressed Stefan‘s comments.

File: .vuepress/.gitignore [hunk 5]
```diff
@@ -2,3 +2,4 @@ addons-*
 public/logos
 thing-types.json
 tmp
+openhab-docs
```
[the remark this answers is not in the document]
Comment (pgfeller) [reply .vuepress/.gitignore:5]: 👍 will remove it from .gitignore in the root folder.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -108,9 +108,23 @@ To run the website on your local machine on a development server with live reloa
 - run the released (stable) version of the documentation execute `npm run build-local-stable -y`
 - run the latest (work in progress) version use `npm run build-local-latest -y`
 
-As the compilation can take a few minutes due to the size of the docs, wait for the "VuePress dev server listening at http://localhost:8080 (or another available port)" message.
+As the compilation can take a few minutes due to the size of the docs, wait for the "VuePress dev server listening at <http://localhost:8080> (or another available port)" message.
 When loading the website in the browser it may take a few seconds until it finally appears.
 
+### Build documentation from PR
+
+In special situations; e.g. if you make changes in the documentation repository that might impact the website build, it is possible to build the website from a PR. This makes it possible to verify that your proposed changes do not have negative side effects:
```
[the remark this answers is not in the document]
Comment (pgfeller) [reply CONTRIBUTING.md:116]: 👍 common mistake of mine; will change this ➡️ unfortunately I did not find a rule yet for the markdown linter for vs code that prevents this in the 1st place.

[the remark this answers is not in the document]
Comment (pgfeller) [reply CONTRIBUTING.md:116]: Btw. I used some `html` to make the page more compact, in case there are no problems with the `ruby` installation. Not sure if we allow the use of such elements. Else we might consider a separate page for such special cases to keep the page compact (not to scare potential contributors away because it looks more complicated than it actually is ...)

File: package.json [hunk 7]
```diff
@@ -28,11 +28,12 @@
   "scripts": {
     "predev": "npm install && npx update-browserslist-db@latest",
     "dev": "npx vuepress dev",
-    "prepare-docs": "ruby prepare-docs.rb",
+    "prepare-docs": "ruby prepare-docs.rb $ARGUMENTS",
     "add-blog-meta": "ruby add-blog-meta.rb",
     "build": "ruby prepare-docs.rb && ruby add-blog-meta.rb && vuepress build .",
     "build-only": "vuepress build .",
-    "build-local": "rm -f javadoc-*.tgz* && npm run prepare-docs && npm run dev",
+    "prebuild-local": "npm run prepare-docs",
```
[the remark this answers is not in the document]
Comment (pgfeller) [reply package.json:35]: Actually the `pre` prefix is a convention of `npm` for a script that is executed automatically as a pre requisit of another step (in that case `build-local`). Therefore it is an internal script that normally is not called by the user.

the `json` specification does not allow comments by default - but most editors can be configured to accept them (as there are many opinionated discussions about this topic). Unfortunately many tools show those scripts as all the others. I can document this - but this is a common mechanism - so I tend to document only the scripts that should be used by the users. I see several options:

- keep as is
- document the magic (that's my least prefered one)
- use `&&` to combine this step with the `build-local` (as it is done in the `build` script) - but if I remember correctly, with this I had problems to pass arguments properly from the command line to the local build setp (therefore also not recommended)
- create a helper script that is called, that will do the necessary steps (but this might be problematic on different platforms - but git bash on windows should be able to handle this) 

Let me know what you think.

File: prepare-docs.rb (hunk 2 above)
Comment (pgfeller) [reply prepare-docs.rb:49]: I'm getting old - and using emoji's helps a bit against this feeling. But I know that probably this is not state of the art anymore with the next generation. But it adds a nice touch to the outputs ... (opinionated - as usual 😉).

File: prepare-docs.rb [hunk 8]
```diff
@@ -5,38 +5,100 @@
...
 
 $ignore_addons = ['transport.modbus', 'transport.feed', 'javasound', 'webaudio', 'oh2']
 
+def verbose(message)
+    if ($verbose == true) then
+      puts message
+    end
+end
+
+def checkout_pull_request(pr, target_directory)
```
[the remark this answers is not in the document]
Comment (pgfeller) [reply prepare-docs.rb:26]: Probably a bug 🐞, or at least unnecessary - I also do not see the need for this ... I'll remove and check if it will still work (it actually should) ...

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (florian-h05) [reply package.json:35]: Hmm seems the last time I worked with npm scripts was too much time ago, now that you say it, I do remember the pre and post scripts.

I would vote for keep as is 👍

File: prepare-docs.rb [hunk 9]
```diff
@@ -348,121 +409,121 @@ def process_file(indir, file, outdir, source)
...
         }
     }
-    puts " -> images"
+    verbose " ->images"
     FileUtils.cp_r(".vuepress/openhab-docs/mainui/images", "docs/mainui")
 
 else
 # Additional files and images for the stable docs
-    puts ">>> Migrating the Settings section"
+    puts "➡️ Migrating the Settings section"
```
[the remark this answers is not in the document]
Comment (pgfeller) [reply prepare-docs.rb:430]: I hope I found all the places; I agree that consistency is important - I hope Florian is ok with the use of unicode characters.

File: prepare-docs.rb (hunk 1 above)
Comment (pgfeller) [reply prepare-docs.rb:8]: Again a personal taste of mine: no dead code in productive things; but rather create a backlog item/issue ... as it was not used I removed it to clean things up a bit. It looks like it is a helper to process HTML & XML files: https://rubygems.org/gems/nokogiri/

I propose to remove it, if it is not used - but I'm also fine to keep it here (of course I would also like to know what the benefit would be).

**Comment by pgfeller** [comment]
I've some pending tasks before a merge can be done:

- [ ] - update to origin/HEAD and resolve conflicts
- [ ] - check, if the feature(s) still works after the changes by the code review & merge/update

File: prepare-docs.rb (hunk 1 above)
Comment (florian-h05) [reply prepare-docs.rb:8]: According to Git blame, `nogokiri` has been replaced by REXML back in 2017, so I'm fine with removing this line.

File: prepare-docs.rb [hunk 10]
```diff
@@ -5,38 +5,99 @@
...
+        $parameter_no_clone = true
+        puts " --no-clone ➡️ existing clone will be used"
+      when "--verbose"
+        $verbose = true
+      else
+        case previous_argument
+            when "--pull-request"
+                $pull_request = arg
+                $version = "final"
+                verbose "ℹ➡️ PR #{$pull_request} will be used to build documentation"
```
Comment (florian-h05) [inline prepare-docs.rb:66]: Why is there an `i`?

File: prepare-docs.rb [hunk 11]
```diff
@@ -558,58 +618,63 @@ def process_file(indir, file, outdir, source)
...
-        puts "    -> images"
+        verbose "    ➡️ images"
         FileUtils.cp_r(".vuepress/openhab-docs/_addons_uis/#{addon}/doc", "addons/ui/#{addon}")
     end
 }
 
+# Git workflow for PRs are note executed - therefore the _addons_ios folder is not present for PR checks - this section will be skipped.
+if $pull_request then
+    puts ""
+    puts "⚠️  Ecosystem documentation for [Alexy, Mycroft, Google] skipped - depends on Git-Workflow ..."
```
Comment (florian-h05) [inline prepare-docs.rb:648]:
```suggestion
    puts "⚠️  Ecosystem documentation for [Alexa, Mycroft, Google] skipped - depends on GitHub action ..."
```

File: CONTRIBUTING.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (florian-h05) [reply CONTRIBUTING.md]: > Not sure if we allow the use of such elements. 

Those elements are allowed and sometimes used in the docs as well :+1:

**Review by florian-h05 — CHANGES_REQUESTED** [review]
Overall looks good to me, I have a few findings during review.

Please have a look at my PR https://github.com/pgfeller/website/pull/1 and merge it.

**Comment by pgfeller** [comment]
@florian-h05 - many thanks for your clean ups 👍

➡️ @stefan-hoehn & @florian-h05 : ... from my side this PR should be ready to merge. I'll have a look into on [#2216](https://github.com/openhab/openhab-docs/issues/2216) next to see if we can use the official footer slot of the template for the edit link to properly use footnotes. I'll also add what markdown extensions we support to the contributor .md

Have a nice (long) weekend & ping me if I missed something in this PR.

@stefan-hoehn - as you may have noticed: I like icons - my favorite console font(s): [Nerd Fonts](https://www.nerdfonts.com)

**Review by florian-h05 — COMMENTED** [review]
LGTM, thanks!!

Just one minor question, and could you please also update https://github.com/openhab/website/blob/fbf16d57b01a38690cb53efd57ddf84b4b45e0dc/add-blog-meta.rb with the arrow icons to make it consistent?

@stefan-hoehn After this PR has been merged I would update the Netlify build setting to to set the verbose argument.

File: package-lock.json
[the remark this answers is not in the document]
Comment (pgfeller) [reply package-lock.json:1]: I checked the file - there are no version changes. I propably deleted the file during development (normaly I do this if I have a build problem) and did not remember to restore the original file. So no need for this change.

**Comment by pgfeller** [comment]
> https://github.com/openhab/website/blob/fbf16d57b01a38690cb53efd57ddf84b4b45e0dc/add-blog-meta.rb with the arrow icons to make it consistent?

Thank you for the review 👍.

File: add-blog-meta.rb [hunk 12]
```diff
@@ -18,7 +18,7 @@
     FileUtils.mkdir_p('.vuepress/tmp')
     FileUtils.mv(file, ".vuepress/tmp/#{File.basename(file)}")
 
-    puts " -> #{file}"
+    puts " ➡️ #{file}"
```
Comment (florian-h05) [inline add-blog-meta.rb:21]:
```suggestion
    puts "   ➡️ #{file}"
```

**Review by florian-h05 — APPROVED** [review]
LGTM now, thanks for the contribution 👍
Just one minor formatting comment left.

@stefan-hoehn Feel free to merge once you have had a look.

**Review by stefan-hoehn — APPROVED** [review]
After talking with @florian-h05 that this has been tested, I approve it now. Thanks to @pgfeller and @florian-h05 for putting so much work and thoughts into it.
